src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs

```diff
@@ -201,6 +201,14 @@ protected override bool TryParseRequest(ReadResult result, out bool endConnectio
         // Suppress pseudo headers from the public headers collection.
         HttpRequestHeaders.ClearPseudoRequestHeaders();
 
+        // Cookies should be merged into a single string separated by "; "
+        var headers = (AspNetCore.Http.IHeaderDictionary)HttpRequestHeaders;
+        if (headers.Cookie.Count > 1)
+        {
+            var mergedCookies = string.Join("; ", headers.Cookie.ToArray());
+            headers.Cookie = new StringValues(mergedCookies);
```

Comment:
Magic: There's an implicit converter from string to StringValues.
```suggestion
            headers.Cookie = string.Join("; ", headers.Cookie.ToArray());
```
https://github.com/dotnet/runtime/blob/2de17e20eaa3ee5ed583fc148bb2ad30e1b33e77/src/libraries/Microsoft.Extensions.Primitives/src/StringValues.cs#L51

Reply:
Also the `header.Cookie.ToArray`-allocation can be avoided. `string.Join` has an overload for IEnumerable, and type-checks the actual implementation.

Reply:
> Also the `header.Cookie.ToArray`-allocation can be avoided. `string.Join` has an overload for IEnumerable, and type-checks the actual implementation.

ToArray doesn't allocate.
https://github.com/dotnet/runtime/blob/2de17e20eaa3ee5ed583fc148bb2ad30e1b33e77/src/libraries/Microsoft.Extensions.Primitives/src/StringValues.cs#L275-L287